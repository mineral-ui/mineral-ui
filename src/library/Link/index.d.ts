declare module '@targetx/mineral-ui/Link';
