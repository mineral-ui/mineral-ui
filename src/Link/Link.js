/**
 * Copyright 2017 CA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */
import React from 'react';
import { createStyledComponent } from '../utils';

type Props = {
  /** Content of the Link */
  children?: MnrlReactNode,
  /** A URL or a URL fragment that the Link points to */
  href?: string,
  /** Element to be used as the root node - e.g. "a" or { ReactRouterLink } */
  element?: $FlowFixMe,
  /** Available variants */
  variant?: 'regular' | 'danger' | 'success' | 'warning' // Should allow string | React class
};

export const componentTheme = (baseTheme: Object) => ({
  Link_borderColor_focus: baseTheme.borderColor_focus,
  Link_color: baseTheme.color_text_primary,
  Link_color_active: baseTheme.color_text_primary_active,
  Link_color_hover: baseTheme.color_text_primary_hover,
  Link_color_focus: baseTheme.color_text_primary_focus,

  ...baseTheme
});

const linkStyles = ({ variant, theme: baseTheme }) => {
  let theme = componentTheme(baseTheme);

  if (variant !== 'regular') {
    // prettier-ignore
    theme = {
      ...theme,
      Link_borderColor_focus: baseTheme[`borderColor_${variant}_focus`],
      Link_color: baseTheme[`color_text_${variant}`],
      Link_color_hover: baseTheme[`color_text_${variant}_hover`],
      Link_color_focus: baseTheme[`color_text_${variant}_focus`],
      Link_color_active: baseTheme[`color_text_${variant}_active`],
    };
  }

  return {
    color: theme.Link_color,
    textDecoration: 'none',

    '&:hover': {
      color: theme.Link_color_hover,
      textDecoration: 'underline'
    },
    '&:focus': {
      color: theme.Link_color_focus,
      outline: `1px solid ${theme.Link_borderColor_focus}`,
      outlineOffset: '2px'
    },
    // `:active` must be last, to follow LVHFA order:
    // https://developer.mozilla.org/en-US/docs/Web/CSS/:active
    '&:active': {
      color: theme.Link_color_active
    }
  };
};

/**
 * The Link component creates a hyperlink to external pages, files, anchors on the same page or another URL.
 * Links trigger an action, select an option, or provide a gateway to more information.
 * Links are used to reduce clutter on a page when there are potentially many actions available.
 *
 * Links with clear messaging and an appropriate variant applied communicate intention effectively.
 */
export default function Link({
  children,
  element = 'a',
  variant = 'regular',
  ...restProps
}: Props) {
  const rootProps = {
    variant,
    ...restProps
  };
  const Root = createStyledComponent(element, linkStyles, {
    displayName: 'Link',
    filterProps: ['variant']
  });

  return (
    <Root {...rootProps}>
      {children}
    </Root>
  );
}
