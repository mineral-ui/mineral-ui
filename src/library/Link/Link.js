/* @flow */
import React from 'react';
import { createStyledComponent } from '../styles';

type Props = {
  /** Content of the Link */
  children?: React$Node,
  /** A URL or URL fragment to which the Link points */
  href?: string,
  /** Element to be used as the root node - e.g. `a` or `ReactRouterLink` */
  element?: $FlowFixMe, // Should allow string | React class
  /** Available variants */
  variant?: 'regular' | 'danger' | 'success' | 'warning'
};

export const componentTheme = (baseTheme: Object) => ({
  Link_borderColor_focus: baseTheme.borderColor_theme_focus,
  Link_color: baseTheme.color_theme,
  Link_color_active: baseTheme.color_theme_active,
  Link_color_hover: baseTheme.color_theme_hover,
  Link_color_focus: baseTheme.color_theme_focus,

  ...baseTheme
});

const linkStyles = ({ variant, theme: baseTheme }) => {
  let theme = componentTheme(baseTheme);

  if (variant !== 'regular') {
    // prettier-ignore
    theme = {
      ...theme,
      Link_borderColor_focus: baseTheme[`borderColor_${variant}_focus`],
      Link_color: baseTheme[`color_${variant}`],
      Link_color_hover: baseTheme[`color_${variant}_hover`],
      Link_color_focus: baseTheme[`color_${variant}_focus`],
      Link_color_active: baseTheme[`color_${variant}_active`],
    };
  }

  return {
    color: theme.Link_color,
    textDecoration: 'none',

    '&:hover': {
      color: theme.Link_color_hover,
      textDecoration: 'underline'
    },
    '&:focus': {
      color: theme.Link_color_focus,
      outline: `1px solid ${theme.Link_borderColor_focus}`,
      outlineOffset: '2px'
    },
    // `:active` must be last, to follow LVHFA order:
    // https://developer.mozilla.org/en-US/docs/Web/CSS/:active
    '&:active': {
      color: theme.Link_color_active
    }
  };
};

/**
 * The Link component creates a hyperlink to external pages, files, anchors on the same page, or another URL.
 */
export default function Link({
  children,
  element = 'a',
  variant = 'regular',
  ...restProps
}: Props) {
  const rootProps = {
    variant,
    ...restProps
  };
  const Root = createStyledComponent(element, linkStyles, {
    displayName: 'Link',
    filterProps: ['variant']
  });

  return <Root {...rootProps}>{children}</Root>;
}
