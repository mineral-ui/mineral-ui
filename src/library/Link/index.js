/* @flow */
export { default } from './Link';
