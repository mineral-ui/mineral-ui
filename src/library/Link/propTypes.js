/* @flow */
import { bool, node, oneOf, string } from 'prop-types';
import { VARIANT } from './constants';

export const linkPropTypes = {
  /* TargetX Custom Props */
  fontSize: string,
  textDecoration: string,
  underline: bool,

  /* Built-In Props */
  children: node,
  href: string,
  variant: oneOf(Object.keys(VARIANT))
};
