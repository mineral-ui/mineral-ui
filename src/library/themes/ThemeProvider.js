/* @flow */
import React, { Children } from 'react';
import { ThemeProvider as EmotionThemeProvider } from 'emotion-theming';
import { canUseDOM } from 'exenv';
import { getDocumentFontSize, rtlAlign } from '../utils';
import EventListener from '../EventListener';
import mineralTheme from './mineralTheme';
import DocumentFontSizeContext from './DocumentFontSizeContext';
import DocumentFontSizeProvider from './DocumentFontSizeProvider';

export type Props = {
  /** Components to which the theme will be applied */
  children?: React$Node,
  /**
   * A shallow object of [theme variables](/theming#common-scenarios-theme-structure)
   * and their values or a function that provides such an object.
   */
  theme?: Object | ((baseTheme?: Object) => Object)
};
type Theme = Object | ((baseTheme?: Object) => Object);

let domLoaded = false;

const documentFontSizeCheck = (documentFontSize) => {
  if (canUseDOM) {
    const actual = parseFloat(
      window.getComputedStyle(document.documentElement).fontSize
    );
    const provided = documentFontSize
      ? getDocumentFontSize(documentFontSize)
      : 'undefined';
    if (!((actual === 16 && !documentFontSize) || actual === provided)) {
      const received =
        typeof documentFontSize === 'string'
          ? `'${documentFontSize}'`
          : documentFontSize === undefined
            ? typeof documentFontSize
            : documentFontSize;
      // prettier-ignore
      throw new Error(
        `[mineral-ui/ThemeProvider]: Expected \`theme\` to contain a \`documentFontSize\` equivalent to the document root's fontSize, '${actual}px' or '${(actual / 16) * 100}%'. Instead got: ${received}`
      );
    }
  }
};

const handleDOMContentLoaded = () => {
  domLoaded = true;
};

const Fragment = ({ children }) => children;

const manipulateTheme = (
  baseTheme?: Object = {},
  theme?: Theme = {},
  parentDocumentFontSize?: number | string
) => {
  let outTheme = typeof theme === 'function' ? theme(baseTheme) : theme;

  if (domLoaded && theme) {
    const { documentFontSize } = outTheme;
    documentFontSizeCheck(
      documentFontSize !== undefined ? documentFontSize : parentDocumentFontSize
    );
  }

  return {
    ...baseTheme,
    ...outTheme
  };
};

/**
 * ThemeProvider provides a theme to the tree of components contained within.
 * See the [theming page](/theming) for more information.
 */
const ThemeProvider = (props: Props) => (
  <DocumentFontSizeContext.Consumer>
    {(parentDocumentFontSize) => {
      const { children, theme } = props;

      const providerProps = {
        parentDocumentFontSize,
        theme
      };

      return (
        <Fragment>
          <DocumentFontSizeProvider {...providerProps}>
            <EmotionThemeProvider
              theme={(baseTheme) =>
                manipulateTheme(baseTheme, theme, parentDocumentFontSize)
              }>
              {Children.only(children)}
            </EmotionThemeProvider>
          </DocumentFontSizeProvider>
          <EventListener
            listeners={[
              {
                target: 'document',
                event: 'DOMContentLoaded',
                handler: handleDOMContentLoaded
              }
            ]}
          />
        </Fragment>
      );
    }}
  </DocumentFontSizeContext.Consumer>
);

ThemeProvider.defaultProps = {
  theme: mineralTheme
};

export default ThemeProvider;
