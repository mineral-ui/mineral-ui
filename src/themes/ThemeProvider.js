/* @flow */
import React from 'react';
import { ThemeProvider as GlamorousThemeProvider } from 'glamorous';
import mineralTheme from './mineralTheme';

type Props = {
  /** Components to which the theme will be applied */
  children?: React$Node,
  /** A shallow object of [theme variables](/theming/#common-scenarios-theme-structure) and their values */
  theme?: Object
};

/**
 * ThemeProvider provides a theme to the tree of components contained within.  See the [theming page](/theming) for more information.
 */
export default function ThemeProvider({
  children,
  theme = mineralTheme
}: Props) {
  return (
    <GlamorousThemeProvider theme={theme}>{children}</GlamorousThemeProvider>
  );
}
