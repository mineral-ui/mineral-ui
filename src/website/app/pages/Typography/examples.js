/* @flow */
import React from 'react';
import { createStyledComponent } from '../../../../library/styles';
import Text from '../../../../library/Text';

type TypeExample = {
  description: string,
  examples: Array<Object>
};

type TypeExamples = Array<TypeExample>;

const styles = {
  list: ({ theme }) => ({
    fontSize: theme.fontSize_ui,
    listStyle: 'square',
    paddingLeft: theme.space_inline_lg
  })
};

const List = createStyledComponent('ul', styles.list);
const examples: TypeExamples = [
  {
    description: `### Long-form Text

Default text and caption styles are intended for sections of content that
provide description and assistance. **Paragraph text** is primarily used for
content sections. **Caption Text** is used to highlight or offset content, such
as inline help or when captioning dashboard content.`,
    examples: [
      {
        content: <Text>Paragraph</Text>,
        value: '<p>, <li>',
        color: 'color_gray_100',
        font: 'Regular (14/0.875)'
      },
      {
        content: <Text appearance="prose">Paragraph (Prose)</Text>,
        value: '<p.prose>',
        color: 'color_gray_100',
        font: 'Regular (16/1)'
      },
      {
        content: <Text appearance="mouse">Paragraph (Mouse)</Text>,
        value: '<p.mouse>',
        color: 'color_gray_100',
        font: 'Regular (11/0.6875)'
      },
      {
        content: (
          <List>
            <li>Alpha</li>
            <li>Bravo</li>
            <li>Charlie</li>
          </List>
        ),
        // the design notes have <ol> here as well,
        // but list-style will override any numbers shown on <ol> list items
        value: '<ul>, <ol>',
        color: 'color_gray_100',
        font: 'Regular (14/0.875)'
      },
      {
        content: <Text element="h5">Caption - Strong</Text>,
        value: '<h5>',
        color: 'color_gray_100',
        font: 'Bold (14/0.875)'
      },
      {
        content: <Text element="h6">Caption - Light</Text>,
        value: '<h6>',
        color: 'color_gray_80',
        font: 'Regular (14/0.875)'
      }
    ]
  },
  {
    description: `### Display Text

Use **Display text** sparingly to make a bold statement, or draw attention to
different sections in an application view. Display text is for short titles and
section divisions—it’s not for long-form content.`,
    examples: [
      {
        content: <Text element="h1">Page Title</Text>,
        value: '<h1>',
        color: 'color_gray_100',
        font: 'ExtraBold (34/2.125)'
      },
      {
        content: <Text element="h2">Large Section Header</Text>,
        value: '<h2>',
        color: 'color_gray_80',
        font: 'Bold (28/1.75)'
      },
      {
        content: <Text element="h3">Medium Section Header</Text>,
        value: '<h3>',
        color: 'color_gray_80',
        font: 'Bold (22/1.375)'
      },
      {
        content: <Text element="h4">Small Section Header</Text>,
        value: '<h4>',
        color: 'color_gray_80',
        font: 'Bold (18/1.125)'
      }
    ]
  }
];

export default examples;
