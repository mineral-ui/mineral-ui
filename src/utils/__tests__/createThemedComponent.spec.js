/**
 * Copyright 2017 CA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */
import React from 'react';
import { mount } from 'enzyme';
import { createThemedComponent } from '../index';
import Link from '../../Link';
import ThemeProvider from '../../ThemeProvider';
import Sample from '../../website/app/demos/utils/components/Sample';
import examples from '../../website/app/demos/utils/examples';
import testDemoExamples from '../../../utils/testDemoExamples';

function mountThemedLink(theme) {
  const ThemedLink = createThemedComponent(Link, theme);

  return mount(
    <ThemeProvider>
      <ThemedLink />
    </ThemeProvider>
  );
}

function mountThemedSample() {
  const ThemedSample = createThemedComponent(Sample, {
    color_primary: 'mediumvioletred'
  });

  return mount(
    <ThemeProvider>
      <ThemedSample />
    </ThemeProvider>
  );
}

describe('createThemedComponent', () => {
  it('renders correctly', () => {
    const themedSample = mountThemedSample();

    expect(themedSample).toMatchSnapshot();
  });

  it('renders a Link correctly with a custom theme object', () => {
    const themedLink = mountThemedLink({
      Link_color: 'mediumvioletred'
    });

    expect(themedLink).toMatchSnapshot();
  });

  it('renders a Link correctly with a custom theme function', () => {
    const themedLink = mountThemedLink(({ theme }) => ({
      Link_color: theme.color_text_danger
    }));

    expect(themedLink).toMatchSnapshot();
  });

  testDemoExamples(examples, { exclude: ['sample-component'] });
});
