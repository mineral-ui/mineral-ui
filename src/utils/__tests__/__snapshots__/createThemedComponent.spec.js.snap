// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createThemedComponent renders correctly 1`] = `
.glamor-0,
[data-glamor-0] {
  background-color: lavender;
  color: mediumvioletred;
  display: inline-block;
  outline: 1px solid #c8d1e0;
  padding: 0.5em;
}

<ThemeProvider>
  <ThemeProvider
    theme={
      Object {
        "backgroundColor_danger": "#db2929",
        "backgroundColor_danger_active": "#c71818",
        "backgroundColor_danger_focus": "#db2929",
        "backgroundColor_danger_hover": "#eb4d4d",
        "backgroundColor_disabled": "#ebeff5",
        "backgroundColor_input_danger": "#fce3e3",
        "backgroundColor_input_success": "#e8fcf2",
        "backgroundColor_input_warning": "#fcf1dc",
        "backgroundColor_link_focus": "#dde3ed",
        "backgroundColor_success": "#0a8f4c",
        "backgroundColor_success_active": "#06783f",
        "backgroundColor_success_focus": "#0a8f4c",
        "backgroundColor_success_hover": "#10a35a",
        "backgroundColor_warning": "#e05b2b",
        "backgroundColor_warning_active": "#cf4615",
        "backgroundColor_warning_focus": "#e05b2b",
        "backgroundColor_warning_hover": "#ed774c",
        "borderColor": "#c8d1e0",
        "borderColor_active": "#51b3f0",
        "borderColor_danger": "#ad0e0e",
        "borderColor_danger_active": "#7d0606",
        "borderColor_danger_focus": "#c71818",
        "borderColor_danger_hover": "#db2929",
        "borderColor_focus": "#0f75b0",
        "borderColor_hover": "#2f9fe0",
        "borderColor_success": "#0a8f4c",
        "borderColor_success_active": "#046132",
        "borderColor_success_focus": "#06783f",
        "borderColor_success_hover": "#1fc06f",
        "borderColor_warning": "#d19411",
        "borderColor_warning_active": "#916504",
        "borderColor_warning_focus": "#cf4615",
        "borderColor_warning_hover": "#e3a322",
        "borderRadius_1": "0.1875em",
        "color_black": "#1d1f24",
        "color_caption": "#58606e",
        "color_gray_10": "#f5f7fa",
        "color_gray_100": "#333840",
        "color_gray_20": "#ebeff5",
        "color_gray_30": "#dde3ed",
        "color_gray_40": "#c8d1e0",
        "color_gray_50": "#afbacc",
        "color_gray_60": "#8e99ab",
        "color_gray_70": "#707a8a",
        "color_gray_80": "#58606e",
        "color_gray_90": "#434a54",
        "color_placeholder": "#8e99ab",
        "color_text": "#333840",
        "color_text_danger": "#ad0e0e",
        "color_text_danger_active": "#940909",
        "color_text_danger_focus": "#ad0e0e",
        "color_text_danger_hover": "#c71818",
        "color_text_disabled": "#afbacc",
        "color_text_ondanger": "#fff",
        "color_text_onprimary": "#fff",
        "color_text_onsuccess": "#fff",
        "color_text_onwarning": "#fff",
        "color_text_primary": "#0f75b0",
        "color_text_primary_active": "#0a6091",
        "color_text_primary_focus": "#0f75b0",
        "color_text_primary_hover": "#1b8bcc",
        "color_text_success": "#06783f",
        "color_text_success_active": "#046132",
        "color_text_success_focus": "#06783f",
        "color_text_success_hover": "#0a8f4c",
        "color_text_warning": "#bd3909",
        "color_text_warning_active": "#a83207",
        "color_text_warning_focus": "#bd3909",
        "color_text_warning_hover": "#cf4615",
        "color_theme_10": "#e1f3fc",
        "color_theme_100": "#084d75",
        "color_theme_20": "#c0e5fc",
        "color_theme_30": "#9fd9fc",
        "color_theme_40": "#79c7f7",
        "color_theme_50": "#51b3f0",
        "color_theme_60": "#2f9fe0",
        "color_theme_70": "#1b8bcc",
        "color_theme_80": "#0f75b0",
        "color_theme_90": "#0a6091",
        "color_white": "#fff",
        "direction": "ltr",
        "fontFamily": "Open Sans",
        "fontFamily_monospace": "monospace",
        "fontFamily_system": "-apple-system, system-ui, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
        "fontSize_base": 16,
        "fontSize_h1": "2.125em",
        "fontSize_h2": "1.75em",
        "fontSize_h3": "1.375em",
        "fontSize_h4": "1.125em",
        "fontSize_h5": "0.875em",
        "fontSize_h6": "0.875em",
        "fontSize_mouse": "0.6875em",
        "fontSize_prose": "1em",
        "fontSize_ui": "0.875em",
        "fontWeight_bold": 700,
        "fontWeight_extraBold": 800,
        "fontWeight_regular": 400,
        "fontWeight_semiBold": 600,
        "lineHeight": 1.25,
        "lineHeight_prose": 1.5,
        "shadow_1": "rgba(0,0,0,0.5) 0 1px 2px",
        "shadow_2": "rgba(0,0,0,0.5) 0 1px 1px, rgba(0,0,0,0.35) 0 2px 4px",
        "shadow_3": "rgba(0,0,0,0.5) 0 1px 1px, rgba(0,0,0,0.25) 0 2px 8px",
        "shadow_4": "rgba(0,0,0,0.5) 0 1px 1px, rgba(0,0,0,0.30) 0 4px 10px",
        "shadow_5": "rgba(0,0,0,0.5) 0 1px 2px, rgba(0,0,0,0.60) 0 15px 15px",
        "size_jumbo": "3.25em",
        "size_large": "2.5em",
        "size_medium": "2em",
        "size_small": "1.5em",
        "spacing_double": "1em",
        "spacing_half": "0.25em",
        "spacing_oneAndAHalf": "0.75em",
        "spacing_quad": "2em",
        "spacing_quarter": "0.125em",
        "spacing_single": "0.5em",
        "spacing_triple": "1.5em",
        "zIndex_100": 100,
        "zIndex_1600": 1600,
        "zIndex_200": 200,
        "zIndex_400": 400,
        "zIndex_800": 800,
      }
    }
  >
    <ThemedComponent>
      <Themed(Sample)
        theme={
          Object {
            "backgroundColor_danger": "#db2929",
            "backgroundColor_danger_active": "#c71818",
            "backgroundColor_danger_focus": "#db2929",
            "backgroundColor_danger_hover": "#eb4d4d",
            "backgroundColor_disabled": "#ebeff5",
            "backgroundColor_input_danger": "#fce3e3",
            "backgroundColor_input_success": "#e8fcf2",
            "backgroundColor_input_warning": "#fcf1dc",
            "backgroundColor_link_focus": "#dde3ed",
            "backgroundColor_success": "#0a8f4c",
            "backgroundColor_success_active": "#06783f",
            "backgroundColor_success_focus": "#0a8f4c",
            "backgroundColor_success_hover": "#10a35a",
            "backgroundColor_warning": "#e05b2b",
            "backgroundColor_warning_active": "#cf4615",
            "backgroundColor_warning_focus": "#e05b2b",
            "backgroundColor_warning_hover": "#ed774c",
            "borderColor": "#c8d1e0",
            "borderColor_active": "#51b3f0",
            "borderColor_danger": "#ad0e0e",
            "borderColor_danger_active": "#7d0606",
            "borderColor_danger_focus": "#c71818",
            "borderColor_danger_hover": "#db2929",
            "borderColor_focus": "#0f75b0",
            "borderColor_hover": "#2f9fe0",
            "borderColor_success": "#0a8f4c",
            "borderColor_success_active": "#046132",
            "borderColor_success_focus": "#06783f",
            "borderColor_success_hover": "#1fc06f",
            "borderColor_warning": "#d19411",
            "borderColor_warning_active": "#916504",
            "borderColor_warning_focus": "#cf4615",
            "borderColor_warning_hover": "#e3a322",
            "borderRadius_1": "0.1875em",
            "color_black": "#1d1f24",
            "color_caption": "#58606e",
            "color_gray_10": "#f5f7fa",
            "color_gray_100": "#333840",
            "color_gray_20": "#ebeff5",
            "color_gray_30": "#dde3ed",
            "color_gray_40": "#c8d1e0",
            "color_gray_50": "#afbacc",
            "color_gray_60": "#8e99ab",
            "color_gray_70": "#707a8a",
            "color_gray_80": "#58606e",
            "color_gray_90": "#434a54",
            "color_placeholder": "#8e99ab",
            "color_text": "#333840",
            "color_text_danger": "#ad0e0e",
            "color_text_danger_active": "#940909",
            "color_text_danger_focus": "#ad0e0e",
            "color_text_danger_hover": "#c71818",
            "color_text_disabled": "#afbacc",
            "color_text_ondanger": "#fff",
            "color_text_onprimary": "#fff",
            "color_text_onsuccess": "#fff",
            "color_text_onwarning": "#fff",
            "color_text_primary": "#0f75b0",
            "color_text_primary_active": "#0a6091",
            "color_text_primary_focus": "#0f75b0",
            "color_text_primary_hover": "#1b8bcc",
            "color_text_success": "#06783f",
            "color_text_success_active": "#046132",
            "color_text_success_focus": "#06783f",
            "color_text_success_hover": "#0a8f4c",
            "color_text_warning": "#bd3909",
            "color_text_warning_active": "#a83207",
            "color_text_warning_focus": "#bd3909",
            "color_text_warning_hover": "#cf4615",
            "color_theme_10": "#e1f3fc",
            "color_theme_100": "#084d75",
            "color_theme_20": "#c0e5fc",
            "color_theme_30": "#9fd9fc",
            "color_theme_40": "#79c7f7",
            "color_theme_50": "#51b3f0",
            "color_theme_60": "#2f9fe0",
            "color_theme_70": "#1b8bcc",
            "color_theme_80": "#0f75b0",
            "color_theme_90": "#0a6091",
            "color_white": "#fff",
            "direction": "ltr",
            "fontFamily": "Open Sans",
            "fontFamily_monospace": "monospace",
            "fontFamily_system": "-apple-system, system-ui, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
            "fontSize_base": 16,
            "fontSize_h1": "2.125em",
            "fontSize_h2": "1.75em",
            "fontSize_h3": "1.375em",
            "fontSize_h4": "1.125em",
            "fontSize_h5": "0.875em",
            "fontSize_h6": "0.875em",
            "fontSize_mouse": "0.6875em",
            "fontSize_prose": "1em",
            "fontSize_ui": "0.875em",
            "fontWeight_bold": 700,
            "fontWeight_extraBold": 800,
            "fontWeight_regular": 400,
            "fontWeight_semiBold": 600,
            "lineHeight": 1.25,
            "lineHeight_prose": 1.5,
            "shadow_1": "rgba(0,0,0,0.5) 0 1px 2px",
            "shadow_2": "rgba(0,0,0,0.5) 0 1px 1px, rgba(0,0,0,0.35) 0 2px 4px",
            "shadow_3": "rgba(0,0,0,0.5) 0 1px 1px, rgba(0,0,0,0.25) 0 2px 8px",
            "shadow_4": "rgba(0,0,0,0.5) 0 1px 1px, rgba(0,0,0,0.30) 0 4px 10px",
            "shadow_5": "rgba(0,0,0,0.5) 0 1px 2px, rgba(0,0,0,0.60) 0 15px 15px",
            "size_jumbo": "3.25em",
            "size_large": "2.5em",
            "size_medium": "2em",
            "size_small": "1.5em",
            "spacing_double": "1em",
            "spacing_half": "0.25em",
            "spacing_oneAndAHalf": "0.75em",
            "spacing_quad": "2em",
            "spacing_quarter": "0.125em",
            "spacing_single": "0.5em",
            "spacing_triple": "1.5em",
            "zIndex_100": 100,
            "zIndex_1600": 1600,
            "zIndex_200": 200,
            "zIndex_400": 400,
            "zIndex_800": 800,
          }
        }
      >
        <ThemeProvider
          theme={
            Object {
              "color_primary": "mediumvioletred",
            }
          }
        >
          <ThemeProvider
            theme={
              Object {
                "color_primary": "mediumvioletred",
              }
            }
          >
            <Sample
              theme={
                Object {
                  "backgroundColor_danger": "#db2929",
                  "backgroundColor_danger_active": "#c71818",
                  "backgroundColor_danger_focus": "#db2929",
                  "backgroundColor_danger_hover": "#eb4d4d",
                  "backgroundColor_disabled": "#ebeff5",
                  "backgroundColor_input_danger": "#fce3e3",
                  "backgroundColor_input_success": "#e8fcf2",
                  "backgroundColor_input_warning": "#fcf1dc",
                  "backgroundColor_link_focus": "#dde3ed",
                  "backgroundColor_success": "#0a8f4c",
                  "backgroundColor_success_active": "#06783f",
                  "backgroundColor_success_focus": "#0a8f4c",
                  "backgroundColor_success_hover": "#10a35a",
                  "backgroundColor_warning": "#e05b2b",
                  "backgroundColor_warning_active": "#cf4615",
                  "backgroundColor_warning_focus": "#e05b2b",
                  "backgroundColor_warning_hover": "#ed774c",
                  "borderColor": "#c8d1e0",
                  "borderColor_active": "#51b3f0",
                  "borderColor_danger": "#ad0e0e",
                  "borderColor_danger_active": "#7d0606",
                  "borderColor_danger_focus": "#c71818",
                  "borderColor_danger_hover": "#db2929",
                  "borderColor_focus": "#0f75b0",
                  "borderColor_hover": "#2f9fe0",
                  "borderColor_success": "#0a8f4c",
                  "borderColor_success_active": "#046132",
                  "borderColor_success_focus": "#06783f",
                  "borderColor_success_hover": "#1fc06f",
                  "borderColor_warning": "#d19411",
                  "borderColor_warning_active": "#916504",
                  "borderColor_warning_focus": "#cf4615",
                  "borderColor_warning_hover": "#e3a322",
                  "borderRadius_1": "0.1875em",
                  "color_black": "#1d1f24",
                  "color_caption": "#58606e",
                  "color_gray_10": "#f5f7fa",
                  "color_gray_100": "#333840",
                  "color_gray_20": "#ebeff5",
                  "color_gray_30": "#dde3ed",
                  "color_gray_40": "#c8d1e0",
                  "color_gray_50": "#afbacc",
                  "color_gray_60": "#8e99ab",
                  "color_gray_70": "#707a8a",
                  "color_gray_80": "#58606e",
                  "color_gray_90": "#434a54",
                  "color_placeholder": "#8e99ab",
                  "color_text": "#333840",
                  "color_text_danger": "#ad0e0e",
                  "color_text_danger_active": "#940909",
                  "color_text_danger_focus": "#ad0e0e",
                  "color_text_danger_hover": "#c71818",
                  "color_text_disabled": "#afbacc",
                  "color_text_ondanger": "#fff",
                  "color_text_onprimary": "#fff",
                  "color_text_onsuccess": "#fff",
                  "color_text_onwarning": "#fff",
                  "color_text_primary": "#0f75b0",
                  "color_text_primary_active": "#0a6091",
                  "color_text_primary_focus": "#0f75b0",
                  "color_text_primary_hover": "#1b8bcc",
                  "color_text_success": "#06783f",
                  "color_text_success_active": "#046132",
                  "color_text_success_focus": "#06783f",
                  "color_text_success_hover": "#0a8f4c",
                  "color_text_warning": "#bd3909",
                  "color_text_warning_active": "#a83207",
                  "color_text_warning_focus": "#bd3909",
                  "color_text_warning_hover": "#cf4615",
                  "color_theme_10": "#e1f3fc",
                  "color_theme_100": "#084d75",
                  "color_theme_20": "#c0e5fc",
                  "color_theme_30": "#9fd9fc",
                  "color_theme_40": "#79c7f7",
                  "color_theme_50": "#51b3f0",
                  "color_theme_60": "#2f9fe0",
                  "color_theme_70": "#1b8bcc",
                  "color_theme_80": "#0f75b0",
                  "color_theme_90": "#0a6091",
                  "color_white": "#fff",
                  "direction": "ltr",
                  "fontFamily": "Open Sans",
                  "fontFamily_monospace": "monospace",
                  "fontFamily_system": "-apple-system, system-ui, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
                  "fontSize_base": 16,
                  "fontSize_h1": "2.125em",
                  "fontSize_h2": "1.75em",
                  "fontSize_h3": "1.375em",
                  "fontSize_h4": "1.125em",
                  "fontSize_h5": "0.875em",
                  "fontSize_h6": "0.875em",
                  "fontSize_mouse": "0.6875em",
                  "fontSize_prose": "1em",
                  "fontSize_ui": "0.875em",
                  "fontWeight_bold": 700,
                  "fontWeight_extraBold": 800,
                  "fontWeight_regular": 400,
                  "fontWeight_semiBold": 600,
                  "lineHeight": 1.25,
                  "lineHeight_prose": 1.5,
                  "shadow_1": "rgba(0,0,0,0.5) 0 1px 2px",
                  "shadow_2": "rgba(0,0,0,0.5) 0 1px 1px, rgba(0,0,0,0.35) 0 2px 4px",
                  "shadow_3": "rgba(0,0,0,0.5) 0 1px 1px, rgba(0,0,0,0.25) 0 2px 8px",
                  "shadow_4": "rgba(0,0,0,0.5) 0 1px 1px, rgba(0,0,0,0.30) 0 4px 10px",
                  "shadow_5": "rgba(0,0,0,0.5) 0 1px 2px, rgba(0,0,0,0.60) 0 15px 15px",
                  "size_jumbo": "3.25em",
                  "size_large": "2.5em",
                  "size_medium": "2em",
                  "size_small": "1.5em",
                  "spacing_double": "1em",
                  "spacing_half": "0.25em",
                  "spacing_oneAndAHalf": "0.75em",
                  "spacing_quad": "2em",
                  "spacing_quarter": "0.125em",
                  "spacing_single": "0.5em",
                  "spacing_triple": "1.5em",
                  "zIndex_100": 100,
                  "zIndex_1600": 1600,
                  "zIndex_200": 200,
                  "zIndex_400": 400,
                  "zIndex_800": 800,
                }
              }
            >
              <glamorous(span)>
                <span
                  className="glamor-0"
                >
                  Sample
                </span>
              </glamorous(span)>
            </Sample>
          </ThemeProvider>
        </ThemeProvider>
      </Themed(Sample)>
    </ThemedComponent>
  </ThemeProvider>
</ThemeProvider>
`;
